gc.c

```diff
@@ -15015,4 +15011,43 @@ RubyUpcalls ruby_upcalls = {
     rb_mmtk_scan_object_ruby_style,
 };
 
+// Use up to 80% of memory for the heap
+static const int rb_mmtk_heap_limit_percentage = 80;
+
+static size_t rb_mmtk_system_physical_memory(void)
+{
+#ifdef __linux__
+    const long physical_pages = sysconf(_SC_PHYS_PAGES);
+    const long page_size = sysconf(_SC_PAGE_SIZE);
+    if (physical_pages == -1 || page_size == -1)
+    {
+        rb_bug("failed to get system physical memory size");
+    }
+    return (size_t) physical_pages * (size_t) page_size;
+#else
+#error no implementation of rb_mmtk_system_physical_memory on this platform
+#endif
+}
+
+static size_t rb_mmtk_available_system_memory(void)
+{
+    /*
+     * If we're in a container, we should use the maximum container memory,
+     * otherwise each container will try to use all system memory. There's
+     * example logic for this in the JVM and SVM (see CgroupV1Subsystem
+     * and CgroupV2Subsystem).
+     */
+
+    return rb_mmtk_system_physical_memory();
+}
+
+size_t rb_mmtk_heap_limit(void) {
+    const char *envval;
+    if ((envval = getenv("THIRD_PARTY_HEAP_LIMIT")) != 0) {
+        return atol(envval);
+    } else {
+        return rb_mmtk_available_system_memory() / rb_mmtk_heap_limit_percentage * 100;
```

Comment:
I think you mean `rb_mmtk_available_system_memory() * rb_mmtk_heap_limit_percentage / 100`.  Otherwise, if `rb_mmtk_heap_limit_percentage == 80`, the return value will be `rb_mmtk_available_system_memory() / 80 * 100` which is `rb_mmtk_available_system_memory() * 1.25`.

```suggestion
        return rb_mmtk_available_system_memory() * rb_mmtk_heap_limit_percentage / 100;
```

Reply:
I spent ages thinking about this and got it wrong!

But I think it should be `system / 100 * percentage`, as your suggested `system * percentage / 100` could overflow couldn't it? (Currently it's `system / percentage * 100`.)

Reply:
Yes if the system has 160 petabyte of physical memory, or 4 megabytes of memory for a 32-bit system.  I think the formula `system / 100 * percentage` you suggested is more reasonable given that it is less likely to have less than 100 bytes of system memory.